feat(api): add gripper grip and release to OT3api

ahiuchingau opened the pull request:
<!--
Thanks for taking the time to open a pull request! Please make sure you've read the "Opening Pull Requests" section of our Contributing Guide:

https://github.com/Opentrons/opentrons/blob/edge/CONTRIBUTING.md#opening-pull-requests

To ensure your code is reviewed quickly and thoroughly, please fill out the sections below to the best of your ability!
-->

# Overview
Adds very basic functions for gripper jaw movements. We can specify a force in Newton and a duration in seconds for the gripper grip and release. Grip would move the gripper jaw inwards to close. Release would move the jaw outwards until the homing switch is triggered.

ahiuchingau committed:
add grip to ot3api

ahiuchingau committed:
implement grip & release

ahiuchingau committed:
fix lint

ahiuchingau committed:
add tests

sfoster1 commented on api/src/opentrons/hardware_control/backends/ot3utils.py:
```diff
@@ -216,6 +218,38 @@ def create_home_group(
     return move_group
 
 
+def create_gripper_move_group(
```
let's call this `create_gripper_jaw_move_group` or something

sfoster1 commented on api/src/opentrons/hardware_control/instruments/gripper_handler.py:
```diff
@@ -57,3 +69,22 @@ def get_gripper_dict(self) -> Optional[GripperDict]:
+        if gripper.has_gripped:
+            raise GripError("Gripper is already gripping")
+
+    def set_ready_to_grip(self, val: bool) -> None:
```
might as well make these properties with setters like
```
@property 
def ready_for_grip(self) -> bool:
    gripper = self._verify_gripper
    return gripper.ready_to_grip

@ready_for_grip.setter
def ready_for_grip(self, ready: bool) -> None:
    gripper = self._verify_gripper()
    gripper.ready_to_grip = ready
```

sfoster1 commented on api/src/opentrons/hardware_control/backends/ot3controller.py:
```diff
@@ -330,6 +332,28 @@ async def fast_home(
         """
         return await self.home(axes)
 
+    async def gripper_move(
```
```suggestion
    async def gripper_move_jaw(
```
?

sfoster1 commented on api/src/opentrons/hardware_control/backends/ot3controller.py:
```diff
@@ -330,6 +332,28 @@ async def fast_home(
+        await runner.run(can_messenger=self._messenger)
+        return True
+
+    async def gripper_home(
```
```suggestion
    async def gripper_home_jaw(
```
?

sfoster1 commented on api/src/opentrons/hardware_control/ot3api.py:
```diff
@@ -902,6 +902,45 @@ async def update_config(self, **kwargs: Any) -> None:
+            self._log.exception("Gripper home failed")
+            raise
+
+    async def prepare_for_grip(self) -> None:
```
I don't think we need this - I think we can happily assert that you have to manually release before gripping.

sfoster1 commented on api/src/opentrons/hardware_control/ot3api.py:
```diff
@@ -902,6 +902,45 @@ async def update_config(self, **kwargs: Any) -> None:
+    async def grip(self, duration: float, newton: float) -> None:
+        dc = self._gripper_handler.get_duty_cycle_by_grip_force(newton)
+        await self._grip(duration=duration, duty_cycle=dc)
+        self._gripper_handler.set_has_gripped(True)
```
wondering if this state is better served by an enum, like
`GRIPPED` (good grip), `RELEASED` (all the way open), `UNHOMED` (haven't homed yet), `CLOSED` (closed but not gripping), `POSITION_CONTROL` (some intermediate value that's not gripped or closed or opened)

sfoster1 reviewed (changes requested):
Some feedback some of which is bikeshedding and some of which isn't. Definitely looking good though

sfoster1 committed:
add _jaw to a bunch of places

sfoster1 committed:
use state enums, integrate into api home

sfoster1 committed:
wrong duty cycle

sfoster1 committed:
more z current

sfoster1 committed:
refactor(hardware): remove gripper freq setting

sfoster1 committed:
duration is not top level

force-limited moves on the gripper jaw will in the future last until we
detect that they have made contact; we do not specify duration.

sfoster1 committed:
wrong frequency

sfoster1 commented:
This needs https://github.com/Opentrons/ot3-firmware/pull/402

sfoster1 commented:
Changes:
- Removing duration; in general, these commands are force-controlled and move until they apply that force or another condition is met. For now, internally they're going for a set duration, but in the future it'll be controlled
- Removing frequency; this is not something we really need to specify, it should be 32khz all the time.

DerekMaggio commented on api/src/opentrons/config/defaults_ot3.py:
```diff
@@ -213,21 +213,21 @@
         OT3AxisKind.X: 1.0,
         OT3AxisKind.Y: 1.0,
-        OT3AxisKind.Z: 1.0,
+        OT3AxisKind.Z: 1.4,
```
Are these defaults found experimentally?

sfoster1 replied:
yeah

DerekMaggio commented on api/src/opentrons/config/gripper_config.py:
```diff
@@ -71,3 +71,28 @@ def load(
+    sub-list in the sequence contains:
+
+    :return: the duty-cycle value for the specified force
+    """
```
Would it be helpful to provide a link to the force/duty-cycle function we are using?

sfoster1 replied:
like, to the file in shared-data, here? https://github.com/Opentrons/opentrons/blob/a7c1796360f5a398a546de4b1887792f78c39fe9/shared-data/gripper/definitions/1/gripperV1.json#L26-L39

DerekMaggio replied:
Yeah, that is helpful. How did we generate these values? Experimentally again?

sfoster1 replied:
Yup

DerekMaggio commented on api/src/opentrons/config/gripper_config.py:
```diff
@@ -71,3 +71,28 @@ def load(
+    """
+    Takes a force in newton and a sequence representing a piecewise
+    function for the slope and y-intercept of a force/duty-cycle function, where each
+    sub-list in the sequence contains:
```
This comment doesn't seem finished. What does each sub-list contain?

DerekMaggio commented on api/src/opentrons/hardware_control/ot3api.py:
```diff
@@ -827,6 +836,8 @@ async def home(
             checked_axes = [ax for ax in OT3Axis]
         async with self._motion_lock:
             try:
+                if OT3Axis.G in checked_axes and self._gripper_handler.has_gripper():
```
Do we modify the OT3Axis enum at Runtime? 
Because looking at [OT3Axis](https://github.com/Opentrons/opentrons/pull/10987/files#diff-3e98a8cae713a483cd2f9b731d8a5c36d5506943c3483bbbf3e6a423f49942f5R115-R124) it looks like `OT3Axis.G` will always exist. If we do not modify the Enum at runtime, I think we can remove the check

sfoster1 replied:
The input handling above is:
- If the user provided `axes`, make sure it becomes a list of OT3Axis elements
- if the user provided nothing, use all the axes

So you could call `api.home([OT3Axis.X, OT3Axis.Y])` and only the X and Y axes would be homed. so line 839 has to account for that.

DerekMaggio replied:
Oh sorry, missed that. Definitely a little rusty after being gone for so long. Thanks for clarifying

DerekMaggio commented on api/src/opentrons/hardware_control/types.py:
```diff
@@ -444,3 +444,21 @@ class NoTipAttachedError(RuntimeError):
+    HOLDING_CLOSED = enum.auto()
+    #: the gripper is in position-control mode somewhere other than its
+    #: open position and probably should be opened before gripping something
+    HOLDING_OPENED = enum.auto()
```
Why wouldn't we always just home the gripper instead of having an intermediate holding open step?

sfoster1 replied:
That's for the future. In this first implementation, we'll definitely be holding the gripper homed. The thing is that the gripper jaw is a DC motor that will need to be actively controlled and will likely jitter a little; so when we add that control loop, we'll want the default position to be like 1mm away from the switch so it doesn't keep bouncing off it.

Laura-Danielle commented on api/src/opentrons/config/defaults_ot3.py:
```diff
@@ -213,21 +213,21 @@
     },
     gripper={
-        OT3AxisKind.Z: 1.0,
+        OT3AxisKind.Z: 1.4,
```
Are we switching to tmc2160 for the gripper? If not, this current is kinda high for the 2130.

sfoster1 replied:
it's pretty high but unfortunately it's necessary, the axis binds up otherwise. Uh we should probably separate the gripper z axis in axis kinds at some point because it's really separate from the other two.

Laura-Danielle commented on api/src/opentrons/hardware_control/backends/ot3utils.py:
```diff
@@ -28,9 +28,15 @@
 )
 
+GRIPPER_JAW_HOME_TIME: float = 120
+GRIPPER_JAW_GRIP_TIME: float = 1
```
Should these default configurations live in shared data at some point?

sfoster1 replied:
No because these are only short term workarounds - the gripper will not be moving for a duration, it will be moving either until it sees that it's reached where it wanted to go, or doing force control.

DerekMaggio reviewed (approved):
LGTM, besides comment stuff in `piecewise_force_conversion`, but that shouldn't stop it from being merged

Laura-Danielle commented on api/src/opentrons/hardware_control/instruments/gripper_handler.py:
```diff
@@ -13,14 +13,29 @@ class GripperNotAttachedError(Exception):
         self._gripper = gripper
 
-    def _verify_gripper(self) -> Gripper:
+    def has_gripper(self) -> bool:
```
Why do we need both `has_gripper` and `verify_gripper`?

sfoster1 replied:
verify_gripper raises, has_gripper doesn't - they're useful in slightly different cases

Laura-Danielle commented on api/src/opentrons/hardware_control/instruments/gripper_handler.py:
```diff
@@ -13,14 +13,29 @@ class GripperNotAttachedError(Exception):
+    def has_gripper(self) -> bool:
+        return bool(self._gripper)
+
+    def verify_gripper(self) -> Gripper:
```
It seems very weird to me to return a gripper object from a function called `verify_gripper`. It doesn't seem like this function should return anything.

sfoster1 replied:
Yeah, what's a better name for it? `get_gripper`?

Laura-Danielle replied:
sure that works I suppose

sfoster1 replied:
https://github.com/Opentrons/opentrons/pull/10987/commits/e1ab52db0ec7d5ba73c85ff83a6eccf8c44b19e9

sfoster1 committed:
comment fixup

sfoster1 committed:
home jaw in backend after other axes

sfoster1 reviewed (approved):
good

Laura-Danielle commented on api/src/opentrons/hardware_control/backends/ot3controller.py:
```diff
@@ -315,6 +319,22 @@ async def fast_home(
+        move_group = create_gripper_jaw_home_group()
+        runner = MoveGroupRunner(move_groups=[move_group])
+        await runner.run(can_messenger=self._messenger)
+        return True
```
I don't see the usage of the boolean anywhere. We should remove it if it's not needed, otherwise we should verify that the message was successful via the can driver.

sfoster1 replied:
https://github.com/Opentrons/opentrons/pull/10987/commits/708c532265cd1706fe6a83a59b1a76327e75a166

Laura-Danielle commented on api/src/opentrons/hardware_control/instruments/gripper_handler.py:
```diff
@@ -13,14 +13,29 @@ class GripperNotAttachedError(Exception):
+
 class GripperHandler:
     def __init__(self, gripper: Optional[Gripper] = None):
         self._gripper = gripper
```
I know I am a bit late to the party on reviewing this refactor, but it really doesn't make sense to me to create/have a gripper handler if no gripper is found in cache instruments. Is that something that will be done in a follow-up?

sfoster1 replied:
Well, it pushes the optional handling down to inside the gripper handler - it's done by analogy with the instrument handler, it's an always-present container for an optionally-present object.

Laura-Danielle replied:
I would argue that there should be a separate 'what instruments are available' class/container to handle that. Especially since this guy has the responsibility of determining state on the gripper itself and grabbing configurations.

But again, something that can be discussed another time and not really relevant to this PR.

Laura-Danielle commented on api/src/opentrons/hardware_control/instruments/gripper_handler.py:
```diff
@@ -57,3 +72,27 @@ def get_gripper_dict(self) -> Optional[GripperDict]:
+    def set_jaw_state(self, state: GripperJawState) -> None:
+        self.verify_gripper().state = state
+
+    def get_duty_cycle_by_grip_force(self, newton: Optional[float] = 0) -> float:
```
Can we make the default `None` here?

sfoster1 replied:
https://github.com/Opentrons/opentrons/pull/10987/commits/5c9d433fa3a38ba841cac4c583414387906b8ace

Laura-Danielle commented on api/src/opentrons/hardware_control/ot3api.py:
```diff
@@ -550,6 +551,14 @@ async def home_z(
+        Home the jaw of the gripper.
+        """
+        gripper = self._gripper_handler.verify_gripper()
+        await self._release()
```
Can we call this `release_gripper`, otherwise it's a bit generic for me. Technically the 96 channel could also "release" tips.

sfoster1 replied:
what about `ungrip`?

Laura-Danielle replied:
sure

sfoster1 replied:
https://github.com/Opentrons/opentrons/pull/10987/commits/3b239614733b5dfc731647b81c3cf6ff0b2f1088

Laura-Danielle commented on hardware/opentrons_hardware/hardware_control/constants.py:
```diff
@@ -1,5 +1,8 @@
 """Hardware Control Constants."""
 from typing_extensions import Final
 
+gripper_interrupts_per_sec: Final = 32000
```
Can we instead call this `brushed_motor_interrupts_per_sec` and `stepper_motor_interrupts_per_sec` since we have a few places where we include `gripper_z` as a way to refer to the Z stepper motor on the gripper?

sfoster1 replied:
> Most of my comments are around naming stuff. I am still a bit confused around the separation of `XHandlerProvider` classes and `XInstrument` classes where X is either pipette or gripper, though I was late to the game on reviewing this stuff. I don't know why we wouldn't ever have a gripper and/or pipette at the time that those objects are created because it seems like something that should be handled by `cache_instruments`.

It admittedly makes a bit less sense with the gripper, but with the pipettes it's because the instrument handler exists to commonize a bunch of code that's the same between the ot2 and ot3 implementations, and it's separate from whether there's a pipette or not. The gripper_handler is by analogy to the pipettes, though there's nothing to hold the implementations in common with. I think it's reasonable to simplify these in the future.

sfoster1 replied:
> Can we instead call this `brushed_motor_interrupts_per_sec` and `stepper_motor_interrupts_per_sec` since we have a few places where we include `gripper_z` as a way to refer to the Z stepper motor on the gripper?

https://github.com/Opentrons/opentrons/pull/10987/commits/27bb0e4ba8cf1495b3b6881ded51716dc4d16cd5

Laura-Danielle reviewed (approved):
Most of my comments are around naming stuff. I am still a bit confused around the separation of `XHandlerProvider` classes and `XInstrument` classes where X is either pipette or gripper, though I was late to the game on reviewing this stuff. I don't know why we wouldn't ever have a gripper and/or pipette at the time that those objects are created because it seems like something that should be handled by `cache_instruments`.

I don't think that those concerns need to be addressed by this PR though, so feel free to only address some of the comments I put.

sfoster1 committed:
brushed_motor_interrupts_per_sec

sfoster1 committed:
gripper controller jaw no bool

sfoster1 committed:
use none as default

sfoster1 committed:
UNGRIP

sfoster1 committed:
get_gripper

Laura-Danielle reviewed (approved):
Changes look good. Let's discuss the instrument handling stuff separately. Def not needed for this PR.
